platform/default/src/mbgl/storage/file_source_manager.cpp

```diff
@@ -0,0 +1,43 @@
+#include <mbgl/storage/asset_file_source.hpp>
+#include <mbgl/storage/database_file_source.hpp>
+#include <mbgl/storage/file_source_manager.hpp>
+#include <mbgl/storage/local_file_source.hpp>
+#include <mbgl/storage/main_resource_loader.hpp>
+#include <mbgl/storage/online_file_source.hpp>
+#include <mbgl/storage/resource_options.hpp>
+
+namespace mbgl {
+
+class DefaultFileSourceManagerImpl final : public FileSourceManager {
+public:
+    DefaultFileSourceManagerImpl() {
+        registerFileSourceFactory(FileSourceType::ResourceLoader, [](const ResourceOptions& options) {
+            return std::make_unique<MainResourceLoader>(options);
+        });
+
+        registerFileSourceFactory(FileSourceType::Asset, [](const ResourceOptions& options) {
+            return std::make_unique<AssetFileSource>(options.assetPath());
+        });
+
+        registerFileSourceFactory(FileSourceType::Database, [](const ResourceOptions& options) {
+            return std::make_unique<DatabaseFileSource>(options);
+        });
+
+        registerFileSourceFactory(FileSourceType::FileSystem,
+                                  [](const ResourceOptions&) { return std::make_unique<LocalFileSource>(); });
+
+        registerFileSourceFactory(FileSourceType::Network, [](const ResourceOptions& options) {
+            auto networkSource = std::make_unique<OnlineFileSource>();
+            networkSource->setProperty("access-token", options.accessToken());
+            networkSource->setProperty("api-base-url", options.baseURL());
+            return networkSource;
+        });
+    }
+};
+
+FileSourceManager* FileSourceManager::get() noexcept {
+    static DefaultFileSourceManagerImpl instance;
+    return &instance;
```

Comment:
If this is non-nullable, why not return a reference insteaf of a pointer?

Reply:
Two reasons:
- API style. Layer manager uses same pattern
- Map don't even need file source manager, for example, when fully inlined style is used (map.cpp)
`FileSourceManager::get() ? FileSourceManager::get()->getFileSource(ResourceLoader, resourceOptions) : nullptr,`